src/test/ui/nll/borrowed-local-error.rs

```diff
@@ -0,0 +1,24 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(nll)]
+
+fn gimme(x: &(u32,)) -> &u32 {
+    &x.0
+}
+
+fn main() {
+    let x = gimme({
+        let v = (22,);
+        &v
+        //~^ ERROR `v` does not live long enough [E0597]
+    });
+    println!("{:?}", x);
```

Comment:
Current output for this test:

```
lunch-box. rustc +nightly borrowed-local-error.rs -Znll-dump-cause
error[E0597]: `v` does not live long enough
  --> borrowed-local-error.rs:20:9
   |
18 |       let x = gimme({
   |  _____________-
19 | |         let v = (22,);
20 | |         &v
   | |         ^^ borrowed value does not live long enough
21 | |         //~^ ERROR `v` does not live long enough [E0597]
22 | |     });
   | |_____-- borrow later used here
   |       |
   |       borrowed value only lives until here

error: aborting due to previous error
```

Reply:
So this output is pretty terrible, but mostly because of the multi-line highlight and the extra text about `must be valid for lifetime '_#2r`.

However, the "borrowed value only lives until here" and "borrow later used here" are also not especially helpful. 


@estebank -- how do you normally deal with those btw? I remember it's a bit of a case-by-case basis; I think in this case the span comes from the MIR. Are there some helper functions we can invoke to try and reduce the span to a single line?

Reply:
This is what I think it *should* look like:

```
error[E0597]: `v` does not live long enough
  --> borrowed-local-error.rs:20:9
     |
  18 |       let x = gimme({
     |               ----- borrow later used here, during the call
  19 |           let v = (22,);
  20 |           &v
     |           ^^ borrowed value does not live long enough
  21 |       });
     |       - borrowed value only lives until here
```

Reply:
I think that to make this happen, we need to:

(a) identify that the later use is a `call` and 'shorten' the highlight to just the function name and/or method name -- @estebank may be able to help with that point

(b) modify the text, since it is due to a function call, to say "during the call"

Reply:
Also:

> borrowed value only lives until here

that is ungreat. I think it should be more of action, like "`v` is dropped here". So let me take another crack at this. I also think the "sequence" here is perhaps a bit confusing, so i'm going to try adding numbers:

```
error[E0597]: `v` does not live long enough
  --> borrowed-local-error.rs:20:9
     |
  18 |       let x = gimme({
     |               ----- (3) But the reference is later used here, during the call
  19 |           let v = (22,);
  20 |           &v
     |           ^^ (1) First, `v` is borrowed here, creating a reference
  21 |       });
     |       - (2) Then, `v` is dropped here, at the end of the block
```

Reply:
I think all of this is feasible: for example, when the message is about something going out of scope, and that thing is a local variable (`v`) rather than a temporary, we can say "at the end of the block", since that is always when local variables go out of scope.

Reply:
Given that the order things happen is meaningful to providing a good explanation, we should either use span notes:

```
error[E0597]: `v` does not live long enough
  --> borrowed-local-error.rs:20:9
     |
  20 |           &v
     |           ^^ borrowed here, creating a reference
  21 |       });
     |       - (2) `v` is dropped here, at the end of the block
     |
note: ...but the reference must live long enough to be used here, during the call
     |
  18 |       let x = gimme({
     |               -----
```

or change the text to be "orderless":

```
error[E0597]: `v` does not live long enough
  --> borrowed-local-error.rs:20:9
     |
  18 |       let x = gimme({
     |               ----- `v` must live long enough to be used in this call, but it doesn't
  19 |           let v = (22,);
  20 |           &v
     |           ^^ borrowed here, creating a reference
  21 |       });
     |       - (2) `v` is dropped here, at the end of the block
     |
```

----

As for wether there's a method to extract the function call, there isn't one now and in all errors where we point at the function call itself we've kept the span around since the AST. I'd have to dig in the current code to see if we do have it available already, but I don't think it should be that hard to find/add.